string.c

```diff
@@ -1874,6 +1874,94 @@ rb_str_init(int argc, VALUE *argv, VALUE str)
     return str;
 }
 
+static size_t
+str_embed_size(long capa)
+{
+    return offsetof(struct RString, as.embed.ary) + (sizeof(char) * capa);
+}
+
+static bool
+str_embeddable_p(long capa)
+{
+    return rb_gc_size_allocatable_p(str_embed_size(capa));
+}
+
+static VALUE
+string_create(rb_execution_context_t *ec, VALUE klass, VALUE orig, VALUE orig_notset, VALUE encoding, VALUE capacity)
+{
+    rb_encoding *enc = NULL;
+    long termlen = 1;
+
+    if (orig_notset == Qtrue) {
+        orig = Qnil;
+    }
+    else {
+        orig = StringValue(orig);
+    }
+
+    if (NIL_P(encoding)) {
+        if (!NIL_P(orig)) {
+            encoding = rb_obj_encoding(orig);
+        }
+    }
+
+    if (!NIL_P(encoding)) {
+        enc = rb_to_encoding(encoding);
+        termlen = rb_enc_mbminlen(enc);
+    }
```

Comment:
Isn't this the same as:
```suggestion
    if (NIL_P(encoding)) {
        if (!NIL_P(orig)) {
            encoding = rb_obj_encoding(orig);
        }
    } else {
        enc = rb_to_encoding(encoding);
        termlen = rb_enc_mbminlen(enc);
    }
```

Reply:
Haha, no.

If `!NIL_P(orig)`, then we'll set `encoding`. It's super confusing I know :)